pkg/collector/corechecks/containers/containerd.go

```diff
@@ -0,0 +1,359 @@
+			}
+		}
+		output.Tags = append(output.Tags, userTags...)
+		output.Ts = e.Timestamp.Unix()
+		output.Title = fmt.Sprintf("Event on %s from Containerd", split[1])
+		if split[1] == "containers" || split[1] == "tasks" {
+			// For task events, we use the container ID in order to query the Tagger's API
+			tags, err := tagger.Tag(e.ID, true)
+			if err != nil {
+				// If there is an error retrieving tags from the Tagger, we can still submit the event as is.
+				log.Errorf("Could not retrieve tags for the container %s: %v", e.ID, err)
+			}
+			output.Tags = append(output.Tags, tags...)
+		}
+		sender.Event(output)
+	}
+}
+
+func computeMetrics(sender aggregator.Sender, nk context.Context, cu cutil.ContainerdItf, userTags []string) {
+	containers, err := cu.Containers()
+	if err != nil {
+		log.Errorf(err.Error())
+		return
+	}
+
+	for _, ctn := range containers {
+		t, errTask := ctn.Task(nk, nil)
+		if errTask != nil {
+			log.Tracef("Could not retrieve metrics from task %s: %s", ctn.ID()[:12], errTask.Error())
+			continue
+		}
+
+		tags, err := collectTags(ctn, nk)
+		if err != nil {
+			log.Errorf("Could not collect tags for container %s: %s", ctn.ID()[:12], err)
+		}
+		tags = append(tags, userTags...)
+		// Tagger tags
+		taggerTags, err := tagger.Tag(ctn.ID(), false)
+		if err != nil {
+			log.Errorf(err.Error())
+			continue
+		}
+		tags = append(tags, taggerTags...)
+
+		metrics, err := convertTasktoMetrics(t, nk)
+		if err != nil {
+			log.Errorf("Could not process the metrics from %s: %v", ctn.ID(), err.Error())
+			continue
+		}
+
+		err = computeExtra(sender, ctn, nk, tags)
+		if err != nil {
+			log.Errorf("Could not process metadata related metrics for %s: %v", ctn.ID()[:12], err.Error())
+		}
+
+		if metrics.Memory.Size() > 0 {
+			computeMem(sender, metrics.Memory, tags)
+		}
+
+		if metrics.CPU.Throttling != nil && metrics.CPU.Usage != nil {
+			computeCPU(sender, metrics.CPU, tags)
+		}
+
+		if metrics.Blkio.Size() > 0 {
+			computeBlkio(sender, metrics.Blkio, tags)
+		}
+
+		if len(metrics.Hugetlb) > 0 {
+			computeHugetlb(sender, metrics.Hugetlb, tags)
+		}
+	}
+}
+
+func convertTasktoMetrics(task containerd.Task, nsCtx context.Context) (*cgroups.Metrics, error) {
+	metricTask, err := task.Metrics(nsCtx)
+	if err != nil {
+		return nil, err
+	}
+
+	metricAny, err := typeurl.UnmarshalAny(&types.Any{
+		TypeUrl: metricTask.Data.TypeUrl,
+		Value:   metricTask.Data.Value,
+	})
+	if err != nil {
+		log.Errorf(err.Error())
+		return nil, err
+	}
+	return metricAny.(*cgroups.Metrics), nil
+}
+
+func computeExtra(sender aggregator.Sender, ctn containerd.Container, nsCtx context.Context, tags []string) error {
+	img, err := ctn.Image(nsCtx)
+	if err != nil {
+		return err
+	}
+	size, err := img.Size(nsCtx)
+	if err != nil {
+		return err
+	}
+	sender.Gauge("containerd.image.size", float64(size), "", tags)
+	return nil
+}
+
+func collectTags(ctn containerd.Container, nsCtx context.Context) ([]string, error) {
```

Comment:
Is this intended to eventually move to the tagger?

Reply:
In a perfect world, we could have a containerd provider yes, but the amount of metadata provided is not worth the time investment at the moment imo